### PR TITLE
Cleanup Job after a configured ttl

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

**Comment by TommyLike** [comment]
@hzxuzhonghu great! Is it for #129 ?

**Comment by hzxuzhonghu** [comment]
Should be merged after #128

File: pkg/apis/batch/v1alpha1/job.go [hunk 1]
```diff
@@ -70,6 +70,15 @@ type JobSpec struct {
...
 	MaxRetry int32 `json:"maxRetry,omitempty" protobuf:"bytes,8,opt,name=maxRetry"`
+
+	// ttlSecondsAfterFinished limits the lifetime of a Job that has finished
+	// execution (either Completed or Failed). If this field is set,
+	// ttlSecondsAfterFinished after the Job finishes, it is eligible to be
+	// automatically deleted. If this field is unset,
+	// the Job won't be automatically deleted. If this field is set to zero,
+	// the Job becomes eligible to be deleted immediately after it finishes.
+	// +optional
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"varint,8,opt,name=ttlSecondsAfterFinished"`
```
Comment (hzxuzhonghu) [inline pkg/apis/batch/v1alpha1/job.go]: Should we set a default value for this ?  Current behavior is when no set, gc will not delete Job.

Comment (TommyLike) [reply pkg/apis/batch/v1alpha1/job.go]: index -> 9

File: pkg/controllers/garbagecollector/garbagecollector.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		return err
+	} else if !expired {
+		return nil
+	}
+
+	// The Job's TTL is assumed to have expired, but the Job TTL might be stale.
+	// Before deleting the Job, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	// The latest Job may have a different UID, but it's fine because the checks will be run again.
+	fresh, err := gb.vkClient.BatchV1alpha1().Jobs(namespace).Get(name, metav1.GetOptions{})
```
Comment (TommyLike) [inline pkg/controllers/garbagecollector/garbagecollector.go]: In which case, we need to do a double check ?

File: pkg/controllers/job/state/aborting.go [hunk 3]
```diff
@@ -31,17 +33,18 @@ func (ps *abortingState) Execute(action vkv1.Action) error {
 		// Already in Restarting phase, just sync it
 		return KillJob(ps.job, func(status *vkv1.JobStatus) {
 			status.State.Phase = vkv1.Restarting
+			status.State.LastTransitionTime = metav1.Now()
```
Comment (TommyLike) [inline pkg/controllers/job/state/aborting.go]: There a lots of codes:
```
status.State.LastTransitionTime = metav1.Now()
```
which can be upgraded into a function which used to update status and update the `LastTransitionTime`  when required.

File: pkg/apis/batch/v1alpha1/job.go (hunk 1 above)
Comment (k82cn) [reply pkg/apis/batch/v1alpha1/job.go]: > Should we set a default value for this ?  Current behavior is when no set, gc will not delete Job.

Yes, we should set default for this in admission controller.

File: pkg/apis/batch/v1alpha1/job.go [hunk 4]
```diff
@@ -218,6 +227,10 @@ type JobState struct {
 	// +optional
 	Phase JobPhase `json:"phase,omitempty" protobuf:"bytes,1,opt,name=phase"`
 
+	// Last time the condition transit from one phase to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
```
Comment (k82cn) [inline pkg/apis/batch/v1alpha1/job.go]: ordered filed by protobuf id

File: pkg/controllers/garbagecollector/garbagecollector.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/controllers/garbagecollector/garbagecollector.go]: We can not guarantee the job TTL not changed between gc received the it and processing here.

Comment (hzxuzhonghu) [reply pkg/controllers/garbagecollector/garbagecollector.go]: Mostly these codes are referred from k8s ttlafterfinished controller

File: pkg/apis/batch/v1alpha1/job.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/apis/batch/v1alpha1/job.go]: > index -> 9

Will update kube-batch

File: pkg/apis/batch/v1alpha1/job.go [hunk 5]
```diff
@@ -70,6 +70,15 @@ type JobSpec struct {
...
 	MaxRetry int32 `json:"maxRetry,omitempty" protobuf:"bytes,8,opt,name=maxRetry"`
+
+	// ttlSecondsAfterFinished limits the lifetime of a Job that has finished
+	// execution (either Completed or Failed). If this field is set,
+	// ttlSecondsAfterFinished after the Job finishes, it is eligible to be
+	// automatically deleted. If this field is unset,
+	// the Job won't be automatically deleted. If this field is set to zero,
+	// the Job becomes eligible to be deleted immediately after it finishes.
+	// +optional
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"varint,9,opt,name=ttlSecondsAfterFinished"`
```
Comment (SrinivasChilveri) [inline pkg/apis/batch/v1alpha1/job.go]: I think we should add the validation as it should not be negative

Comment (hzxuzhonghu) [reply pkg/apis/batch/v1alpha1/job.go]: validate can be applied in admission controller

Comment (SrinivasChilveri) [reply pkg/apis/batch/v1alpha1/job.go]: > validate can be applied in admission controller

yeah thats correct we should add the validation in admission controller

Comment (hzxuzhonghu) [reply pkg/apis/batch/v1alpha1/job.go]: IIRC @wangyuqing4 will work on admission controller.

**Comment by hzxuzhonghu** [comment]
@k82cn This is ready to go?

File: pkg/apis/batch/v1alpha1/job.go (hunk 5 above)
Comment (k82cn) [reply pkg/apis/batch/v1alpha1/job.go]: @hzxuzhonghu , please confirm with  @wangyuqing4 ; if she is not working this, please feel free to open a PR for that.

**Pull request merged**